File: .API/TransactionManager.cs
// Decompiled with JetBrains decompiler
// Type: CloudX.Shared.TransactionManager
// Assembly: CloudX.Shared, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
// MVID: 2635624C-5F24-4EFB-ACD1-7E9C1349B2F5
// Assembly location: F:\SteamLibrary\steamapps\common\NeosVR\HeadlessClient\CloudX.Shared.dll

using BaseX;
using System;
using System.Threading.Tasks;

namespace CloudX.Shared
{
  public class TransactionManager
  {
    public CloudXInterface Cloud { get; private set; }

    public Decimal? NCRConversionRatio { get; private set; }

    public TransactionManager(CloudXInterface cloud)
    {
      this.Cloud = cloud;
      Task.Run((Func<Task>) (async () => await this.LoadConversionData()));
    }

    private async Task LoadConversionData()
    {
      CloudResult<Decimal> cloudResult = await this.Cloud.ReadGlobalVariable<Decimal>(TransactionUtil.NCR_CONVERSION_VARIABLE);
      if (cloudResult.IsOK)
        this.NCRConversionRatio = new Decimal?(cloudResult.Entity);
      else
        UniLog.Warning("Error getting conversion ratio. " + cloudResult.State.ToString() + "\n\n" + cloudResult.Content, false);
    }

    public Decimal? TryConvert(string sourceToken, Decimal sourceAmount, string targetToken)
    {
      if (sourceToken == "USD")
      {
        if (targetToken == null || !(targetToken == "NCR"))
          return new Decimal?();
        Decimal num = sourceAmount;
        Decimal? ncrConversionRatio = this.NCRConversionRatio;
        if (!ncrConversionRatio.HasValue)
          return new Decimal?();
        return new Decimal?(num / ncrConversionRatio.GetValueOrDefault());
      }
      if (!(targetToken == "USD"))
        return new Decimal?();
      if (sourceToken != null)
      {
        if (!(sourceToken == "NCR"))
        {
          if (sourceToken == "KFC")
            return new Decimal?(new Decimal());
        }
        else
        {
          Decimal num = sourceAmount;
          Decimal? ncrConversionRatio = this.NCRConversionRatio;
          if (!ncrConversionRatio.HasValue)
            return new Decimal?();
          return new Decimal?(num * ncrConversionRatio.GetValueOrDefault());
        }
      }
      return new Decimal?();
    }

    public bool IsValidToken(string token)
    {
      return token != null && (token == "NCR" || token == "KFC");
    }

    public Decimal? ToUSD(string token, Decimal amount)
    {
      if (token != null)
      {
        if (!(token == "NCR"))
        {
          if (token == "KFC")
            return new Decimal?(new Decimal());
        }
        else
        {
          if (!this.NCRConversionRatio.HasValue)
            return new Decimal?();
          return new Decimal?(this.NCRConversionRatio.Value * amount);
        }
      }
      throw new Exception("Invalid token: " + token);
    }

    public static string FormatCurrency(Decimal? amount)
    {
      if (!amount.HasValue)
        return "N/A";
      return amount.Value.ToString("#,0.00####################");
    }
  }
}

File: API/TransactionManager.js
const Decimal = require("decimal.js");
const { TransactionUtil } = require("./TransactionUtil");
class TransactionManager {
	constructor(cloud) {
		this.CDFTConversionRatio = null;
		this.NCRConversionRatio = null;
		this.TransactionManager(cloud);
	}
	TransactionManager(cloud) {
		this.Cloud = cloud;
		(async () => {
			await this.LoadConversionData();
		})();
	}
	async LoadConversionData() {
		let cloudResult1 = await this.Cloud.ReadGlobalVariable(
			TransactionUtil.NCR_CONVERSION_VARIABLE
		);
		if (cloudResult1.IsOK) {
			this.NCRConversionRatio = new Decimal(cloudResult1.Entity.value);
		} else {
			throw new Error(
				"Error getting conversion ratio. " +
					cloudResult1.State +
					"\n\n" +
					cloudResult1.Content
			);
		}
		let cloudResult2 = await this.Cloud.ReadGlobalVariable(
			TransactionUtil.CDFT_CONVERSION_VARIABLE
		);
		if (cloudResult2.IsOK) {
			this.CDFTConversionRatio = new Decimal(cloudResult2.Entity.value);
		} else {
			throw new Error(
				"Error getting conversion ratio. " +
					cloudResult2.State.ToString() +
					"\n\n" +
					cloudResult2.Content
			);
		}
	}
	TryConvert(sourceToken, sourceAmount, targetToken) {
		let num1;
		let num2;
		let num3;
		let num4;
		let ncrConversionRatio1;
		let cdftConversionRatio1;
		let ncrConversionRatio2;
		let cdftConversionRatio2;
		if (sourceToken === "USD") {
			switch (targetToken) {
			case "NCR":
				num1 = sourceAmount;
				ncrConversionRatio1 = this.NCRConversionRatio;
				return !(ncrConversionRatio1 != null)
					? new Decimal()
					: num1 / ncrConversionRatio1;
			case "CDFT":
				num2 = sourceAmount;
				cdftConversionRatio1 = this.CDFTConversionRatio;
				return !(cdftConversionRatio1 != null)
					? new Decimal()
					: num2 / cdftConversionRatio1;
			default:
				return new Number();
			}
		} else {
			if (!(targetToken === "USD")) return new Number();
			switch (sourceAmount) {
			case "NCR":
				num3 = sourceAmount;
				ncrConversionRatio2 = this.NCRConversionRatio;
				return !(ncrConversionRatio2 != null)
					? new Decimal()
					: num3 * ncrConversionRatio2;
			case "CDFT":
				num4 = sourceAmount;
				cdftConversionRatio2 = this.CDFTConversionRatio;
				return !(cdftConversionRatio2 != null)
					? new Decimal()
					: num4 * cdftConversionRatio2;
			case "KFC":
				return new Number();
			default:
				return new Number();
			}
		}
	}
	IsValidToken(token) {
		switch (token) {
		case "NCR":
		case "CDFT":
		case "KFC":
			return true;
		default:
			return false;
		}
	}
	ToUSD(token, amount) {
		var cdftConversionRatio;
		var num;
		switch (token) {
		case "NCR":
			return !(this.NCRConversionRatio != null)
				? new Decimal()
				: this.NCRConversionRatio * amount;
		case "CDFT":
			cdftConversionRatio = this.CDFTConversionRatio;
			num = amount;
			return !(cdftConversionRatio != null)
				? new Decimal()
				: cdftConversionRatio * num;
		case "KFC":
			return new Number();
		default:
			throw new Error("Invalid Token: " + token);
		}
	}
	static FormatCurrency(amount) {
		if (!amount) return "N/A";
		return amount.toString();
	}
}
module.exports = {
	TransactionManager,
};
